core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1054,6 +1054,23 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
     schedule.initial.flatMap(loop(None, _))
   }
 
+  /**
+   * Repeats this effect until the partial function succeeds on the result.
+   */
+  final def repeatUntil[R1 <: R, E1 >: E, B](f: PartialFunction[A, ZIO[R1, E1, B]]): ZIO[R1, Nothing, B] =
+    self.flatMap { a =>
+      f.lift(a) match {
+        case Some(value) => value
+        case None        => repeatUntil(f)
+      }
+    }.eventually
+
+  /**
+   * Effectful version of `repeatUntil`. Will repeat until `f` succeeds
+   */
+  final def repeatUntilM[R1 <: R, E1 >: E, B](f: A => ZIO[R1, E1, B]): ZIO[R1, Nothing, B] =
```

Comment:
Sorry for the confusion, I think this one should be `PartialFunction[A, ZIO[R1, E1, B]]`, and that `repeatUntil` should be `PartialFunction[A, B]`. Then there is symmetry between them and it matches the overall ZIO convention on the `M` suffix. Does that make sense?